packages/flutter/lib/src/material/circle_avatar.dart

```diff
@@ -85,48 +93,113 @@ class CircleAvatar extends StatelessWidget {
+      minDiameter = 2.0 * (minRadius ?? _defaultMinRadius);
+    }
+    return minDiameter;
+  }
+
+  double get _maxDiameter {
+    double maxDiameter = _defaultRadius * 2.0;
+    if (radius != null) {
+      maxDiameter = radius * 2.0;
+    } else if (minRadius != null || maxRadius != null) {
+      maxDiameter = 2.0 * (maxRadius ?? _defaultMaxRadius);
+    }
+    return maxDiameter;
+  }
+
   @override
   Widget build(BuildContext context) {
     assert(debugCheckHasMediaQuery(context));
     final ThemeData theme = Theme.of(context);
-    TextStyle textStyle = theme.primaryTextTheme.title;
-    if (foregroundColor != null) {
-      textStyle = textStyle.copyWith(color: foregroundColor);
-    } else if (backgroundColor != null) {
+    TextStyle textStyle = theme.primaryTextTheme.title.copyWith(color: foregroundColor);
+    Color effectiveBackgroundColor = backgroundColor;
+    if (effectiveBackgroundColor != null) {
       switch (ThemeData.estimateBrightnessForColor(backgroundColor)) {
         case Brightness.dark:
-          textStyle = textStyle.copyWith(color: Colors.white);
+          textStyle = textStyle.copyWith(color: theme.primaryColorLight);
           break;
         case Brightness.light:
-          textStyle = textStyle.copyWith(color: Colors.black);
+          textStyle = textStyle.copyWith(color: theme.primaryColorDark);
+          break;
+      }
+    } else {
+      switch (ThemeData.estimateBrightnessForColor(textStyle.color)) {
+        case Brightness.dark:
+          effectiveBackgroundColor = theme.primaryColorLight;
+          break;
+        case Brightness.light:
+          effectiveBackgroundColor = theme.primaryColorDark;
           break;
       }
     }
+    final double minDiameter = _minDiameter;
+    final double maxDiameter = _maxDiameter;
     return new AnimatedContainer(
-      width: radius * 2.0,
-      height: radius * 2.0,
+      constraints: new BoxConstraints(
+        minHeight: minDiameter,
+        minWidth: minDiameter,
+        maxWidth: maxDiameter,
+        maxHeight: maxDiameter,
+      ),
       duration: kThemeChangeDuration,
       decoration: new BoxDecoration(
-        color: backgroundColor ?? theme.primaryColor,
-        image: backgroundImage != null ? new DecorationImage(
-          image: backgroundImage
-        ) : null,
+        color: effectiveBackgroundColor,
+        image: backgroundImage != null ? new DecorationImage(image: backgroundImage) : null,
         shape: BoxShape.circle,
       ),
-      child: child != null ? new Center(
-        child: new MediaQuery(
-          // Need to reset the textScaleFactor here so that the
-          // text doesn't escape the avatar when the textScaleFactor is large.
-          data: MediaQuery.of(context).copyWith(textScaleFactor: 1.0),
-          child: new DefaultTextStyle(
-            style: textStyle.copyWith(color: foregroundColor),
-            child: child,
-          ),
-        )
-      ) : null,
+      child: child != null
```

Comment:
I think this kind of thing is slightly easier to read if you put the child == null so return null case first. Like
```
child: child == null
  ? null,
  : new Center(
      ...
    )
```

Reply:
Yes, I see what you mean.  Fixed.